lua/nvim-treesitter/textobjects.lua

```diff
@@ -66,13 +66,146 @@ function M.select_textobject(query_string)
     else
-      ts_utils.update_selection(bufnr, smallest_range.node)
+      api.nvim_win_set_cursor(api.nvim_get_current_win(),
+                              { adjacent_textobject_range[3] + 1, adjacent_textobject_range[4] })
     end
   end
 end
 
+-- luacheck: push ignore 631
+M.goto_next_textobject_start = function(query_string) M.goto_adjacent_textobeject(query_string, 'forward', 'start', false) end
+M.goto_next_textobject_end = function(query_string) M.goto_adjacent_textobeject(query_string, 'forward', false, false) end
+M.goto_previous_textobject_start = function(query_string) M.goto_adjacent_textobeject(query_string, false, 'start', false) end
+M.goto_previous_textobject_end = function(query_string) M.goto_adjacent_textobeject(query_string, false, false, false) end
+-- luacheck: pop
+
+function M.goto_next_textobject_end(query_string)
+  local bufnr, _, node = get_textobject_at_point(query_string)
+  if not node then return end
+  local next_textobject = M.next_textobject(node,  query_string, false, bufnr)
+  local next_textobject_range = next_textobject:range()
+  api.nvim_win_set_cursor(api.nvim_get_current_win(), { next_textobject_range[3] + 1, next_textobject_range[4] + 1 })
+end
+
+function M.next_textobject(node, query_string, same_parent, bufnr)
+  local node = node or ts_utils.get_node_at_cursor()
+  local bufnr = bufnr or api.nvim_get_current_buf()
+
+  local matches = queries.get_capture_matches(bufnr, query_string, 'textobjects')
+  local _, _ , node_end = node:end_()
+  local next_node
+  local next_node_start
+
+  for _, m in pairs(matches) do
+    local _, _, other_end = m.node:start()
+    if other_end > node_end then
+      if not same_parent or node:parent() == m.node:parent() then
+        if not next_node then
+          next_node = m
+          _, _, next_node_start = next_node.node:start()
+        end
+        if other_end < next_node_start then
+          next_node = m
+          _, _, next_node_start = next_node.node:start()
+        end
+      end
+    end
+  end
+
+  return next_node and next_node.node
+end
+
+function M.previous_textobject(node, query_string, same_parent, bufnr)
+  local node = node or ts_utils.get_node_at_cursor()
+  local bufnr = bufnr or api.nvim_get_current_buf()
+
+  local matches = queries.get_capture_matches(bufnr, query_string, 'textobjects')
+  local _, _ , node_start = node:start()
+  local previous_node
+  local previous_node_end
+
+  for _, m in pairs(matches) do
+    local _, _, other_end = m.node:end_()
+    if other_end < node_start then
+      if not same_parent or node:parent() == m.node:parent() then
+        if not previous_node then
+          previous_node = m
+          _, _, previous_node_end = previous_node.node:end_()
+        end
+        if other_end > previous_node_end then
+          previous_node = m
+          _, _, previous_node_end = previous_node.node:end_()
+        end
+      end
+    end
+  end
```

Comment:
Can't these be refactored a bit ? Using something like a closure to find the appropriate node ?

Reply:
But we have nothing like filter or sort... Lua is copy paste programming. Well we could add the minimum for a more functional programming.

Reply:
We could add a generic function to filter and sort nodes.

Reply:
Is the new version better. At least it could be reused in more situations.

Reply:
As we saw in #252, many languages require a new predicate that allows excluding certain node types (here we need to exclude comments).

So adding @parameter should be separate PR.

Reply:
Yep, looks better.